### PR TITLE
MGMT-2424: Support for LSO(Local Storage Operator) installation

### PR DESCRIPTION
**Pull request by priyanka19-98**
Code for LSO installation in assisted-service 
Signed off by: Priyanka Jiandani<pjiandan@redhat.com>

### CODE REVIEW

File: internal/operators/lso/utils.go [hunk 1]
```diff
@@ -0,0 +1,352 @@
+package lso
+
+func Manifests() map[string]string {
+	manifests := make(map[string]string)
+	manifests["99_openshift-lso_ns.yaml"] = localStorageNamespace
+	manifests["99_openshift-lso_operator_group.yaml"] = lsoOperatorGroup
+	manifests["99_openshift-lso_subscription.yaml"] = lsoSubscription
+	manifests["99_openshift-lso_lvset_cr.yaml"] = localVolumeSet
+	manifests["99_openshift-lso_lvset_crd.yaml"] = localVolumeSetCrd
+	manifests["99_openshift-lso_csv.yaml"] = lsoCatalogSource
```
Comment (rollandf) [inline internal/operators/lso/utils.go:10]: LSO is available in 4.6 channel. Some of these yamls are not needed.

Check here : https://github.com/openshift/assisted-service/pull/571/files#diff-bc992597facc49eaa3e041cfc231e095377671318f9bdf948012167084fa6f64

File: swagger.yaml [hunk 2]
```diff
@@ -1,6 +1,6 @@
-swagger: '2.0'
+swagger: "2.0"
```
Comment (rollandf) [inline swagger.yaml:1]: A lot of unrelated changes.
Please move to a different PR if you think it is needed.

**Review by rollandf — CHANGES_REQUESTED** [review]

**Comment by rollandf** [comment]
@avishayt PTAL at API

File: internal/bminventory/inventory.go [hunk 3]
```diff
@@ -1589,6 +1598,16 @@ func (b *bareMetalInventory) updateClusterData(ctx context.Context, cluster *com
 	if err = network.VerifyClusterCIDRsNotOverlap(machineCidr, clusterCidr, serviceCidr, userManagedNetworking); err != nil {
 		return common.NewApiError(http.StatusBadRequest, err)
 	}
+
+	if params.ClusterUpdateParams.Operators != nil && params.ClusterUpdateParams.Operators.Lso != nil {
+		updates["operators"] = *params.ClusterUpdateParams.Operators
+		updates["operators_lso"] = *params.ClusterUpdateParams.Operators.Lso
+		updates["operators_lso_enabled"] = *params.ClusterUpdateParams.Operators.Lso.Enabled
```
Comment (filanov) [inline internal/bminventory/inventory.go:1605]: Don't you need just this part in the updates list?

File: internal/ignition/ignition.go [hunk 4]
```diff
@@ -140,6 +141,23 @@ func (g *installerGenerator) Generate(ctx context.Context, installConfig []byte)
 		}
 	}
 
+	if *g.cluster.Operators.Lso.Enabled {
```
Comment (filanov) [inline internal/ignition/ignition.go:144]: `Operators` and `Lso` is nullable so i'm not sure that you can access `Enabled` directly without checking that the other parts are not `nil`

Comment (filanov) [inline internal/ignition/ignition.go:144]: consider splitting this code into another function.

**Review by filanov — COMMENTED** [review]
@avishayt  can you please review the swagger?

File: internal/ignition/ignition.go (hunk 4 above)
Comment (rollandf) [reply internal/ignition/ignition.go:144]: I think that in RegisterCluster default values are set, so that should be OK.

File: swagger.yaml [hunk 5]
```diff
@@ -2911,6 +2925,10 @@ definitions:
         type: string
         description: NTP source going to be added to all the hosts.
         x-nullable: true
+      operators:
+        $ref: '#/definitions/operators'
+        x-go-custom-tag: gorm:"embedded;embedded_prefix:operators_"
```
Comment (filanov) [inline swagger.yaml:2930]: it's not stored in the DB so you don't need db specifications here.

File: swagger.yaml [hunk 6]
```diff
@@ -3033,6 +3051,10 @@ definitions:
         type: string
         description: NTP source going to be added to all the hosts.
         x-nullable: true
+      operators:
+        $ref: "#/definitions/operators"
+        x-go-custom-tag: gorm:"embedded;embedded_prefix:operators_"
```
Comment (filanov) [inline swagger.yaml:3056]: it's not stored in the DB so you don't need db specifications here.

File: internal/ignition/ignition.go [hunk 7]
```diff
@@ -140,6 +159,13 @@ func (g *installerGenerator) Generate(ctx context.Context, installConfig []byte)
 		}
 	}
 
+	if g.cluster.Operators != nil && g.cluster.Operators.Lso != nil && *g.cluster.Operators.Lso.Enabled {
```
Comment (filanov) [inline internal/ignition/ignition.go:162]: not tests for ignition?

File: internal/operators/lso/utils.go [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+	return manifests
+}
+
+const lsoSubscription = `apiVersion: operators.coreos.com/v1alpha1
+kind: Subscription
+metadata:
+  name: local-storage-operator
+  namespace: openshift-local-storage
+spec:
+  channel: "4.6"
```
Comment (mhrivnak) [inline internal/operators/lso/utils.go:19]: What is the plan for managing this over time? Particularly in a future where multiple OCP releases are supported for installation?

Comment (rollandf) [reply internal/operators/lso/utils.go:19]: We can pass the OpenShift version to be installed from the Cluster and update the channel accordingly.
If the CRDs changes between versions, then the Manifests() method should handle the supported versions and return the right CRDs.

File: internal/ignition/ignition.go (hunk 7 above)
Comment (rollandf) [reply internal/ignition/ignition.go:162]: Maybe move that logic inside `generateLsoManifests` so that you could add test to it more easily?

File: internal/bminventory/inventory.go (hunk 3 above)
Comment (rollandf) [reply internal/bminventory/inventory.go:1596]: @priyanka19-98 please fix accordingly

File: internal/operators/lso/utils.go (hunk 8 above)
Comment (filanov) [reply internal/operators/lso/utils.go:19]: Cluster have a OCP version so we can probably have this as template, but if there are changes don't include only parameters then we will need to have a map of configurations.

File: internal/bminventory/inventory.go (hunk 3 above)
Comment (priyanka19-98) [reply internal/bminventory/inventory.go:1596]: @filanov , any other changes to be made in this part?

Comment (filanov) [reply internal/bminventory/inventory.go:1596]: i don't think that the db schema have thise:
```
		updates["operators"] = *params.ClusterUpdateParams.Operators
		updates["operators_lso"] = *params.ClusterUpdateParams.Operators.Lso
```

Comment (priyanka19-98) [reply internal/bminventory/inventory.go:1594]: Made the changes

File: internal/ignition/ignition.go (hunk 7 above)
Comment (priyanka19-98) [reply internal/ignition/ignition.go]: Added the tests!

**Review by filanov — APPROVED** [review]
